binsync/common/ui/force_push/panels/functions_table.py

```diff
@@ -0,0 +1,483 @@
+            return self.checkState(QPersistentModelIndex(index))
+        return None
+
+    def headerData(self, section, orientation, role=Qt.DisplayRole):
+        """ Set the headers to be displayed. """
+        if role != Qt.DisplayRole:
+            return None
+
+        if orientation == Qt.Horizontal:
+            if 0 <= section < len(self.HEADER):
+                return self.HEADER[section]
+
+        return None
+
+    def insertRows(self, position, rows=1, index=QModelIndex()):
+        """ Insert N (default=1) rows into the model at a desired position. """
+        self.beginInsertRows(QModelIndex(), position, position + rows - 1)
+
+        for row in range(rows):
+            self.row_data.insert(position + row, [0, "LOADING", "USER", datetime.now()])
+            if self.load_from == "bs": self.data_bgcolors.insert(position + row, [QColor(0, 0, 0, 0)])
+
+        self.endInsertRows()
+        return True
+
+    def removeRows(self, position, rows=1, index=QModelIndex()):
+        """ Remove N (default=1) rows from the model at a desired position. """
+        if 0 <= position < len(self.row_data) and 0 <= position + rows < len(self.row_data):
+            self.beginRemoveRows(QModelIndex(), position, position + rows - 1)
+
+            del self.row_data[position:position + rows]
+            if self.load_from == "bs": del self.colordata[position:position + rows]
+
+            self.endRemoveRows()
+            return True
+        return False
+
+    def setData(self, index, value, role=Qt.EditRole):
+        """ Adjust the data (set it to <value>) depending on the given
+            index and role.
+        """
+        if role != Qt.EditRole and role != Qt.CheckStateRole:
+            return False
+
+        if index.isValid() and 0 <= index.row() < len(self.row_data):
+            address = self.row_data[index.row()]
+            if 0 == index.column() and role == Qt.CheckStateRole:
+                self.checks[QPersistentModelIndex(index)] = value 
+            elif 0 <= index.column() < len(address):
+                address[index.column()] = value
+            else:
+                return False
+            # TODO: Check for compatibility issues with pyqt6 here, function prototype changes between versions
+            self.dataChanged.emit(index, index)
+            return True
+
+        return False
+
+    def flags(self, index):
+        """ Set the item flags at the given index. """
+        if not index.isValid():
+            return Qt.ItemIsEnabled
+        if self.load_from == "decompiler" and index.column()==0: fl = Qt.ItemIsUserCheckable | Qt.ItemIsSelectable | Qt.ItemIsEnabled
+        else: fl = Qt.ItemFlags(QAbstractTableModel.flags(self, index))
+        return fl
+
+    def entry_exists(self, addr):
+        """ Quick way to determine if an entry already exists via addr """
+        return addr in [i[0] for i in self.row_data]
```

Comment:
Hilariously, this function in the _slowest_ way possible to check if an entry already exists. You want to be very careful of stuff like this. You call this function `n` times where `n` is the number of possible functions in the binary. Then inside the function you do a `n` iteration construction to make the list, which you then use `addr in list` to search... which is again another `n` search. In this function alone you are doing a `O(2n)` complexity. Considering where it's called, you are doing a `O(2(n ** 2))`... which is really bad considering this is a UI and a freezing event. 

Here is the correct solution:
1. Whenever you find yourself needing to check the existence of an element in a list, convert it to a `set`. Internally in Python the use of `in` will now be `O(1)`, since its treated like a dict. 
2. At the beginning of the use of this function, create the full `set` once, not every time its called. This will reduce your solution of `O(2(n ** 2))` => `O(2n)` since it will take `n` to create the initial set, and we will do a `O(1)` search `n` times after that.